packages/mwp-router/src/AsyncRoute.jsx

```diff
@@ -10,90 +10,86 @@ type Props = {
+	component,
+	_componentCache: { ...state._componentCache, [key]: component },
+});
 
 /**
- * Route rendering component that will render nested routes asynchronously
- * The nested routes are cached so that the async data is not re-requested
- * each time a route is re-rendered.
+ * Route rendering component that uses internal state to keep a reference to the
+ * wrapping component. If the `route` prop has a statically defined `component`,
+ * it will be used on first render, otherwise it will be resolved using the
+ * component 'getter' and the the result will be rendered (and cached).
  */
 class AsyncRoute extends React.Component<Props, State> {
 	constructor(props: Props) {
 		super(props);
-		const { match, route } = props;
-		const childRoutes = getChildRoutes({ match, route });
+		const route = props.route;
+
 		this.state = {
-			childRoutes,
-			_routesCache: {},
+			component: route.component || PassThrough,
+			_componentCache: {},
 		};
 	}
-	updateChildRoutes(childRoutes: Array<PlatformRoute>, key: string) {
-		this.setState(state => ({
-			childRoutes,
-			_routesCache: {
-				...state._routesCache,
-				[key]: childRoutes,
-			},
-		}));
-	}
-	resolveAsyncChildRoutes(
-		resolver: () => Promise<Array<PlatformRoute> | PlatformRoute>
-	) {
+	/*
+	 * Given a component-resolving function, update this.state.component with
+	 * the resolved value - set/get cached reference as necessary
+	 */
+	resolveComponent(resolver: () => Promise<React$ComponentType<*>>) {
+		this.setState(state => ({ component: PassThrough }));
 		const key = resolver.toString();
-		// async route - check for cache keyed by stringified load function
-		const cached = this.state._routesCache[key];
+		const cached = this.state._componentCache[key];
 		if (cached) {
-			this.updateChildRoutes(cached, key);
+			this.setState({ component: cached });
 			return;
 		}
-		resolver().then(routes => {
-			routes = routes instanceof Array ? routes : [routes];
-			this.updateChildRoutes(routes, key);
-		});
-		return;
+		resolver()
+			.then(getComponentStateSetter(key))
+			.then(setter => this.setState(setter));
 	}
+
 	/*
-	 * New props may correspond to a route change. If so, this function populates
-	 * this.state.childRoutes to correspond to the indexRoute or nested routes
-	 * defined by the new route
+	 * New props may correspond to a route change. If so, this function sets the
+	 * component to render
 	 */
 	componentWillReceiveProps(nextProps: Props) {
 		const { match, route } = nextProps;
 		if (route === this.props.route && match === this.props.match) {
-			return;
-		}
-		const childRoutes = getChildRoutes({ match, route });
-		this.setState(state => ({ childRoutes }));
-		if (childRoutes.length) {
-			return;
+			return; // no new route, just re-render normally
 		}
-		if (match.isExact && route.getIndexRoute) {
-			this.resolveAsyncChildRoutes(route.getIndexRoute);
+
+		if (route.component) {
+			this.setState(state => ({ component: route.component }));
```

Comment:
is it possible that `route.component` would already equal `state.component` ?

Reply:
if two different routes specified the same component, yes. Since new routes will trigger a re-render regardless, however, this isn't 'expensive' in that case.